src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs

```diff
@@ -151,7 +152,18 @@ protected override byte[] DecryptImpl(byte* pbCiphertext, uint cbCiphertext, byt
                         cbDerivedKey: _symmetricAlgorithmSubkeyLengthInBytes);
 
                     // Perform the decryption operation
-
+#if NETCOREAPP
+                    unsafe
+                    {
+                        var nonce = new Span<byte>(pbNonce, (int)NONCE_SIZE_IN_BYTES);
+                        var key = new Span<byte>(pbSymmetricDecryptionSubkey, (int)_symmetricAlgorithmSubkeyLengthInBytes);
+                        var tag = new Span<byte>(pbAuthTag, (int)TAG_SIZE_IN_BYTES);
+                        var plaintext = new Span<byte>(retVal);
+                        var encrypted = new Span<byte>(pbEncryptedData, (int)cbPlaintext);
+                        using var aes = new AesGcm(key);
```

Comment:
This code doesn't belong here. The `Cng` namespace should consist solely of p/invoke calls into BCRYPT. This probably belongs under the `Managed` namespace instead.

Reply:
Ah, so basically all the places that are doing GCM need to have this switch instead of putting it in the Cng/Gcm, and that we should just use the ManagedEncrypter instead?

Reply:
Kinda. There are four implementations of authenticated encryption in the DataProtection stack today. (See [the _ConfigurationModel_ namespace](https://github.com/dotnet/aspnetcore/tree/main/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel).)

1. `AuthenticatedEncryptorDescriptor`, which is a generalized "I tell you the algorithm, you choose the best platform implementation" descriptor.
2. `CngCbcAuthenticatedEncryptorDescriptor`, which is encrypt-then-MAC, specifically using Windows CNG.
3. `CngGcmAuthenticatedEncryptorDescriptor`, which is GCM, specifically using Windows CNG.
4. `ManagedAuthenticatedEncryptorDescriptor`, which is encrypt-then-MAC, specifically using .NET `SymmetricAlgorithm` and `KeyedHashAlgorithm` abstractions.

(2) and (3) are specifically tied to CNG. They allow callers to fulfill compliance requirements by selecting specific CNG implementations, including hardware-based implementations. (Take a look at all of the `[ApplyPolicy]` attributes [here](https://github.com/dotnet/aspnetcore/blob/dfe625fbd122557cc4dda0f3d6f3dc5c7afc8c8e/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel/CngGcmAuthenticatedEncryptorConfiguration.cs).) These code paths should never do anything other than p/invoke into bcrypt.dll.

(4) is generalized in that it uses existing .NET abstractions, but it's encrypt-then-MAC, which isn't relevant to GCM. So let's ignore it for now.

(1) is the most useful for this scenario since it's already generalized. All you need to do is specify the desired algorithm family, and it'll choose the best implementation available on the current operating system. [It already handles GCM.](https://github.com/dotnet/aspnetcore/blob/460d4d95da28d1be8ecc6418cf58a124723e8043/src/DataProtection/DataProtection/src/AuthenticatedEncryption/ConfigurationModel/AuthenticatedEncryptorDescriptor.cs#L41)

In addition to the _ConfigurationModel_ namespace mentioned previously, each of the above descriptors has an associated factory in [the _AuthenticatedEncryption_ namespace](https://github.com/dotnet/aspnetcore/tree/main/src/DataProtection/DataProtection/src/AuthenticatedEncryption). This is where you need to make your targeted change. Remove the OS check from the block below, then create a new internal type which implements _IOptimizedAuthenticatedEncryptor_ and which delegates to a backing `AesGcm` instance.

https://github.com/dotnet/aspnetcore/blob/460d4d95da28d1be8ecc6418cf58a124723e8043/src/DataProtection/DataProtection/src/AuthenticatedEncryption/AuthenticatedEncryptorFactory.cs#L55-L72

In theory that's all you need to do. Maybe with an `#ifdef` around the OS version check if you're targeting a platform where the `AesGcm` type doesn't exist. Then somebody can use [the existing builder function](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/DataProtection/DataProtection/src/DataProtectionBuilderExtensions.cs#L602) and it should work regardless of what platform you're running on.

```cs
builder.UseCryptographicAlgorithms(new AuthenticatedEncryptorConfiguration {
    EncryptionAlgorithm = EncryptionAlgorithm.AES_256_GCM
});
```

Reply:
Thanks @GrabYourPitchforks that information helps a lot, so I started looking working on a new `ManagedGcmAuthenticatedEncryptor` class, the main question I had was whether I needed to still do any of the logic in the existing `GcmAuthenciatedEncryptor` specifically the AesGcm apis require a nonce, and that's obviously not something in the outer Encrypt/Decrypt apis.  So I can generate a random nonce on Encrypt, but it seems like I need to then mimic the assumptions in GcmAuthenticatedEncryptor's DecryptImpl, namely the cipher payload is: 

`keyModifer || nonce || encryptedData || authenticationTag`

Just checking that my understanding is correct?

Reply:
Yes, your understanding is correct. The implementations should be near-identical, except that one uses _AesGcm_ and the other uses CNG. In fact, the two implementations should have 100% compatible payloads. If you create a ManagedGcmAuthenticatedEncryptor initialized over (some_secret, AES_256_GCM), and if you then feed (some_purpose, some_plaintext) to it, the resulting ciphertext should be readable by a CngGcmAuthenticationEncryptor initialized over (same_secret, "AES", "Microsoft default CNG provider") \[or whatever the default provider name is\] and passed (same_purpose, ciphertext) as input; and vice versa.

The ManagedCbcMacAuthenticatedEncryptor and CngCbcMacAuthenticatedEncryptor behave similarly: they can decipher each other's payloads because they create the context header in the same way and they generate the ciphertexts in the same way.

[We even document how the context header is generated](https://docs.microsoft.com/aspnet/core/security/data-protection/implementation/context-headers) in an effort to help formalize this guarantee.

Reply:
As a follow-up, I'd recommend renaming the internal _GcmAuthenticatedEncryptor_ type to _CngGcmAuthenticatedEncryptor_ (to highlight that it's CNG-specific), and I'd name the new internal encryptor type _AesGcmAuthenticatedEncryptor_ (to highlight that it's `AesCng`-specific).

Reply:
Okay I think I precomputed the 3 context headers now :)  I'm looking at the decrypt function and I'm sort of stuck on what the managed replacement for the derive key with context header call would be as I see this calls down into BCrypt's DeriveKey:

https://github.com/dotnet/aspnetcore/blob/b1a59e8c70e53dbe1784ddac6a9fea55565fa56c/src/DataProtection/DataProtection/src/Cng/GcmAuthenticatedEncryptor.cs#L145

Reply:
Check out the type *ManagedSP800_108_CTR_HMACSHA512*, which is a managed implementation of the SP800-108 algorithm. You should be able to call that routine from the new encryptor type. The _prfFactory_ parameter should be the lambda function `(kdk) => new HMACSHA512(kdk)`.

Reply:
Aside: this code is all from the pre-span days. It'd be a welcome refactoring in a separate PR to start spanifying this to avoid all the unsafe goop.

Reply:
Okay, I added the new class and did my best to mimic the existing logic but using the style of the managed encryptor.  The round trip tests are failing that are using the CngAes -> AesGcm round trip tests, and as far as I can tell the surrounding code looks fine, I verified that the keymodifier/nonce/tags are all being split up the same and passed into AesGcm, but the decrypt is failing with an authentication tag does not match.  The subkey derivation looks fine as well, in the debugger in that the same sub key is derived in Cng and Aes.  Maybe it will be obvious to you how I'm clearly holding things wrong slightly?

Reply:
Okay I found it finally, I had a bogus "-" in one of the precomputed context headers which caused the roundtrip between cng and aes to fail :(  The tests are green now including the new Aes round trip variant, so I think this is ready for review now